Parse triple quoted string annotations as if parenthesized

Glyphack opened the pull request:
<!--
Thank you for contributing to Ruff! To help us out with reviewing, please consider the following:

- Does this pull request include a summary of the change? (See below.)
- Does this pull request include a descriptive title?
- Does this pull request include references to any relevant issues?
-->

## Summary

<!-- What's the purpose of the change? What does it do, and why? -->

Resolves #9467 

Parse quoted annotations as if the string content is inside parenthesis.
With this logic `x` and `y` in this example are equal:

```python
y: """
   int |
   str
"""

z: """(
    int |
    str
)
"""
```

Also this rule only applies to triple quotes([link](https://github.com/python/typing-council/issues/9#issuecomment-1890808610)).

This PR is based on the [comments](https://github.com/astral-sh/ruff/issues/9467#issuecomment-2579180991) on the issue.

I did one extra change, since we don't want any indentation tokens I am setting the `State::Other` as the initial state of the Lexer.

Remaining work:

- [x] Add a test case for red-knot.
- [x] Add more tests.

## Test Plan

Added a test which previously failed because quoted annotation contained indentation.
Added an mdtest for red-knot.
Updated previous test.

<!-- How was it tested? -->

Glyphack commented on crates/ruff_python_parser/src/lib.rs:
```diff
@@ -166,6 +168,46 @@ pub fn parse_expression_range(
+    parsed.try_into_expression().unwrap().into_result()
+}
+
+pub fn parse_as_string_annotation(
```
I don't know if this function is necessary. I was worried someone might update the linter parsing logic for string annotations and forget red knot. With this at least it will be applied for both.

carljm commented on crates/red_knot_python_semantic/resources/mdtest/annotations/string.md:
```diff
@@ -173,3 +173,27 @@ p: "call()"
+       str
+  """,
+):
+    reveal_type(a1)  # revealed: None
+    reveal_type(a2)  # revealed: None
+    reveal_type(a3)  # revealed: int | str
+    reveal_type(a4)  # revealed: int | str
```
I assume out of these four, only `a4` failed before this PR? (Can't check right now, since GH git ops are down.)

Not sure what `a1` and `a2` are really testing?

Glyphack replied:
Yes only a4 fails. I added the other ones just to have a single line triple quoted in the tests but I guess it's not really needed. Will remove it.

carljm reviewed (commented):
Thank you!! The red-knot test and changes look fine to me, but that's a small part of this PR. Would rather have @dhruvmanila take a look at the parser/ruff side.

MichaReiser commented on crates/ruff_python_parser/src/lib.rs:
```diff
@@ -599,6 +641,8 @@ pub enum Mode {
     /// [Automatic parentheses and quotes]: https://ipython.readthedocs.io/en/stable/interactive/reference.html#automatic-parentheses-and-quotes
     Ipython,
+
+    ParenthesizedExpression,
```
I'd move it next to `Expression` and let's add a comment explaining what this mode is.

MichaReiser reviewed (commented):
Thank you. Would you mind adding a test where an expression is incorrectly parenthesized? 

```python
def f(
	a: """
		int | 
str)
"""
)
```

or 

```python
def f(
	a: """
		int) | 
str
"""
)
```

I'm curious how it recovers in those cases and if we need to do something more than just setting `nesting = 1`

Glyphack commented:
@MichaReiser I added the test cases you suggested and the tests pass(see this [commit](https://github.com/astral-sh/ruff/pull/15387/commits/b3cf440dfdc9500afd4826b23d97363928cb9ec4))
But I think that is because the parser is recovering.

So I fixed the nesting check to check for nesting value of 1 when mode is `ParenthesizedExpression`. This way the lexer will detect the problem that nesting is one level lower than started as well.

dhruvmanila commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -84,15 +84,22 @@ impl<'src> Lexer<'src> {
             "Lexer only supports files with a size up to 4GB"
         );
 
+        let nesting = u32::from(mode == Mode::ParenthesizedExpression);
+        let state = if mode == Mode::ParenthesizedExpression {
+            State::Other
+        } else {
+            State::AfterNewline
+        };
```
nit:
```suggestion
        let (state, nesting) = if mode == Mode::ParenthesizedExpression {
            (State::Other, 1)
        } else {
            (State::AfterNewline, 0)
        };
```

dhruvmanila commented on crates/ruff_python_parser/src/lib.rs:
```diff
@@ -582,6 +624,11 @@ pub enum Mode {
     /// The code consists of a single expression.
     Expression,
 
+    /// The code is parsed as if it is parenthesized. The code does not need to contain parenthesis
+    /// at the beginning. And it is assumed by the lexer that there is a parenthesis.
+    /// This allows the lexer to allow certain indentations that are not valid in a python file.
+    ParenthesizedExpression,
```
I've suggested some improvements and I don't think we should specify the internal details about implementation within the lexer.

```suggestion
    /// The code consists of a single expression and is parsed as if it is parenthesized. The parentheses themselves aren't required.
    /// This allows for having valid multiline expression without the need of parentheses
    /// and is specifically useful for parsing string annotations.
    ParenthesizedExpression,
```

dhruvmanila commented on crates/ruff_python_parser/src/lib.rs:
```diff
@@ -166,6 +168,46 @@ pub fn parse_expression_range(
+    parsed.try_into_expression().unwrap().into_result()
+}
+
+pub fn parse_as_string_annotation(
```
What do you think about `parse_string_annotation` or `parse_string_annotation_content`? Additionally, we should provide documentation similar to what you've done in `parse_parenthesized_expression_range` as this is a public function and the crate is being used by multiple projects.

dhruvmanila commented on crates/ruff_linter/resources/test/fixtures/pyflakes/F722.py:
```diff
@@ -11,3 +11,30 @@ def g() -> "///":
+)
+"""
+
+# single quotes are not implicitly parenthesized
+invalid: "\n int"
```
Can we add the same test case using triple-quoted string which will be useful as a documentation?

```py
# but, using the same annotation inside a triple-quoted string is valid
valid: """\n int"""
```

Glyphack replied:
Actually I made a mistake here, invalid: "\n int" is in a python file. So this string will not actually have a newline character but only \ and n as characters. So also """"\n int""" fails.
Is there a way to actually create a type annotation that contains newline inside a python file without using triple quotes?

Glyphack replied:
I removed this right now. Since I could not figure out how to do it correctly.

dhruvmanila replied:
I see, that makes sense. I don't think it's even possible to have a newline in a single-quoted string without the `\n` character. This will also utilize `parse_complex_annotation` and not `parse_simple_annotation`.

dhruvmanila commented on crates/ruff_linter/resources/test/fixtures/pyflakes/F722.py:
```diff
@@ -11,3 +11,30 @@ def g() -> "///":
+invalid: "\n int"
+
+
+invalid2: """
+		int |
+str)
+"""
+
+invalid3: """
+		int) |
+str
+"""
```
Can we add some test cases for unclosed parenthesis similar to how these? It would also be useful to provide test cases with multiple parentheses.

dhruvmanila commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -1309,10 +1316,25 @@ impl<'src> Lexer<'src> {
+            // For Mode::ParenthesizedExpression we start with nesting level 1.
+            // So we check if we end with that level.
+            Mode::ParenthesizedExpression => {
+                if self.nesting != 1 {
+                    // Reset the nesting to avoid going into infinite loop.
+                    self.nesting = 1;
+                    return self
+                        .push_error(LexicalError::new(LexicalErrorType::Eof, self.token_range()));
+                }
```
Should this be `if self.nesting > 1`?

Glyphack replied:
I applied the suggestion below and it resolves it.

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -1309,10 +1316,25 @@ impl<'src> Lexer<'src> {
-            // Reset the nesting to avoid going into infinite loop.
-            self.nesting = 0;
-            return self.push_error(LexicalError::new(LexicalErrorType::Eof, self.token_range()));
+        match self.mode {
+            // For Mode::ParenthesizedExpression we start with nesting level 1.
+            // So we check if we end with that level.
+            Mode::ParenthesizedExpression => {
+                if self.nesting != 1 {
+                    // Reset the nesting to avoid going into infinite loop.
+                    self.nesting = 1;
+                    return self
+                        .push_error(LexicalError::new(LexicalErrorType::Eof, self.token_range()));
+                }
+            }
+            _ => {
+                if self.nesting > 0 {
+                    // Reset the nesting to avoid going into infinite loop.
+                    self.nesting = 0;
+                    return self
+                        .push_error(LexicalError::new(LexicalErrorType::Eof, self.token_range()));
+                }
+            }
```
```suggestion
				// For Mode::ParenthesizedExpression we start with nesting level 1.
        // So we check if we end with that level.
				let init_nesting = if self.mode == Mode::ParenthesizedExpression {
					1
				} else {
					0
				};

        if self.nesting > init_nesting {
            // Reset the nesting to avoid going into infinite loop.
            self.nesting = 0;
            return self
                .push_error(LexicalError::new(LexicalErrorType::Eof, self.token_range()));
        }
            
```

MichaReiser reviewed (approved):
Thanks. This looks great.

dhruvmanila commented on crates/ruff_linter/src/rules/pyflakes/snapshots/ruff_linter__rules__pyflakes__tests__F722_F722.py.snap:
```diff
@@ -13,4 +13,68 @@ F722.py:13:4: F722 Syntax error in forward annotation: `List[int]☃`
+15 | # Type annotations with triple quotes can contain newlines and indentation
+   |
+
+F722.py:30:11: F722 Syntax error in forward annotation: `
+		int |
+str)
+`
```
Unrelated to this PR, I think it would be useful to provide the syntax error that was raised during parsing this string content.

dhruvmanila reviewed (approved):
Thank you for working on this!
